### PR TITLE
For discussion - refactored LuxFloat to use P8[],I8[],D8[] and give same output as MwRewrite

### PR DESCRIPTION
**Pull request by martinbudden**
`pidLuxFloat` is refactored to use `P8[]`, `I8[]` and `D8[]` etc, coefficients rather than `P_f[]`, `I_f[]`, `D_f[]` etc.

`PTerm`, `ITerm`, and `DTerm` output is scaled to produce the same output as `pidMultiWiiRewrite`.

Test code is written to check equivalence. (Although if this refactor finds favour I will write more test code.)

This means the same bank of PIDs is used for `pidLuxFloat` and `pidMultiWiiRewrite` and users should be able to swap between `pidLuxFloat` and `pidMultiWiiRewrite` without retuning.

Because `P_f[]`, `I_F[]` etc no longer used simplification of `serial_msp.c`, `serial_cli.c` and `rc_adjustments.c` is possible.

`pidLuxFloat` uses floating point calculations based around the value of `dT`. `pidMultiWiiRewrite` does integer calculations based around the value of `targetLoopTime`, the conversion factors between these two values are defined in `pid_luxfloat.c` as;

```
// constants to scale pidLuxFloat so output is same as pidMultiWiiRewrite
static const float luxPTermScale = 1.0f / 128;
static const float luxITermScale = (1000000.0f / (0x1000000));
static const float luxDTermScale = 3 * (0.000001f * (float)0xFFFF) / 1028; // the 3 is because mwrewrite sums 3 deltas rather than taking moving average
static const float luxGyroScale = 16.4f / 4.0f; // the 16.4 is needed because mwrewrite does not scale according to the gyro model
```

### CODE REVIEW

**Comment by borisbstyle** [comment]
I definately do like this idea.

I will have to go through it better when I am on my PC, but I assume the rates are still the same or did you scale those too to be the same?

**Comment by martinbudden** [comment]
@borisbstyle , I've changed the rate calculation as well. The main reason I did this was so that the test code could directly compare the P, I, and T terms calculated by `pidLuxFloat` and `pidMultiWiiRewrite`. Now it's proven to work I can change them back again (and alter the test code accordingly), if that is the consensus.

**Comment by hydra** [comment]
set to 1.13.0 since 1.13.0 already has loads of cleanup work done. it makes sense.

**Comment by hydra** [comment]
@borisbstyle thoughts regarding @martinbudden 's comments regarding rates?

**Comment by borisbstyle** [comment]
@hydra @martinbudden 
I personally would like same rates and Pids for those pid controllers. That helps with determing defaults!

It's basically a choice. Having equal rates will require user to reconfigure their rates. If we provide enough info how to convert their old rates it should be fine.

btw I would also consider increasing the scaling for rewrite yaw rate to be closer to roll and pitch like current luxfloat

**Comment by ledvinap** [comment]
It would be nice to separate conversion from PidProfile for PID parameters.
Define structure with parameters  applicable to given PID (using floating point vars for Lux) and function that will fill this structure with data from pidProfile (and possibly from other sources).
 - all scaling factors will be localized in this function (luxPTermScale etc)
 - it will be easier to test
 - fine-tuning / experimenting will be easier
 - it will document what variables are used in PID

File: src/main/flight/pid_luxfloat.c [hunk 1]
```diff
@@ -54,147 +54,149 @@
...
+                // multiplication of rcCommand corresponds to changing the sticks scaling here
 #ifdef GPS
-             const int32_t errorAngle = (constrain(rcCommand[axis] + GPS_angle[axis], -((int) max_angle_inclination),
-                    +max_angle_inclination) - attitude.raw[axis] + angleTrim->raw[axis]) / 10.0f; // 16 bits is ok here
+                const float errorAngle = constrain(2 * rcCommand[axis] + GPS_angle[axis], -((int)max_angle_inclination), max_angle_inclination)
+                        - attitude.raw[axis] + angleTrim->raw[axis];
 #else
-             const int32_t errorAngle = (constrain(rcCommand[axis], -((int) max_angle_inclination),
-                    +max_angle_inclination) - attitude.raw[axis] + angleTrim->raw[axis]) / 10.0f; // 16 bits is ok here
+                const float errorAngle = constrain(2 * rcCommand[axis], -((int)max_angle_inclination), max_angle_inclination)
```
Comment (borisbstyle) [inline src/main/flight/pid_luxfloat.c]: Whats the multiplier by 2?

**Comment by borisbstyle** [comment]
Also that averaging can go into function as well

File: src/main/flight/pid_luxfloat.c [hunk 2]
```diff
@@ -54,147 +54,149 @@
...
-            delta1[axis] = delta;
-            delta = deltaSum / 3.0f;
-        }
-
-        DTerm = constrainf(delta * pidProfile->D_f[axis] * PIDweight[axis] / 100, -PID_LUX_FLOAT_MAX_D, PID_LUX_FLOAT_MAX_D);
-
-        // -----calculate total PID output
-        axisPID[axis] = constrain(lrintf(PTerm + ITerm + DTerm), -PID_LUX_FLOAT_MAX_PID, PID_LUX_FLOAT_MAX_PID);
-
+        const float gyroRate = luxGyroScale * gyroADC[axis] * gyro.scale;
```
Comment (borisbstyle) [inline src/main/flight/pid_luxfloat.c]: luxGyroScale is already divided by 4.....so why would you do this?
gyroRate = luxGyroScale * gyroADC[axis] * gyro.scale
can just be
gyroRate = gyroADC[axis] / 4   <------same as rewrite

Comment (borisbstyle) [inline src/main/flight/pid_luxfloat.c]: With your method there is a few % difference

File: src/main/flight/pid_luxfloat.c (hunk 1 above)
Comment (martinbudden) [inline src/main/flight/pid_luxfloat.c]: @borisbstyle , the `2*` multiplier is there to max luxfloat the same as mwrewrite, and mwrewrite has the `2*` multiplier. Why the original mwrewrite has the multiplier and the original luxfloat does not, I don't know.

**Comment by martinbudden** [comment]
@borisbstyle , @hydra , @ledvinap , for reference here are the original Cleanflight rates, my changes, and for comparison, the betaflight rates.

I agree with Boris, in that it's probably best if LUX_FLOAT and MWREWRITE have the save rates.

Any views on what the actual values should be?

```
Current LUX_FLOAT (changed to be same as MWREWRITE)
YAW
angleRate = (float)((rate + 27) * rcCommand[YAW]) / 32.0f;
ROLL, PITCH
angleRate = (float)((rate + 27) * rcCommand[axis]) / 16.0f;

Current MWREWRITE (unchanged)
YAW
angleRate = (((int32_t)(rate + 27) * rcCommand[YAW]) >> 5);
ROLL, PITCH
angleRate = ((int32_t)(rate + 27) * rcCommand[axis]) >> 4;
```
```
Previous LUX_FLOAT
YAW
AngleRate = (float)((rate + 10) * rcCommand[YAW]) / 50.0f;
ROLL, PITCH
AngleRate = (float)((rate + 20) * rcCommand[axis]) / 50.0f;

Previous MWREWRITE
YAW
AngleRateTmp = (((int32_t)(rate + 27) * rcCommand[YAW]) >> 5);
ROLL, PITCH
AngleRateTmp = ((int32_t)(rate + 27) * rcCommand[axis]) >> 4;
```
```
BetaFlight LUX_FLOAT
YAW
AngleRate = (float)((rate + 47) * rcCommand[YAW]) / 32.0f;
ROLL, PITCH
AngleRate = (float)((rate + 27) * rcCommand[axis]) / 16.0f;
   
BetaFlight MWREWRITE
YAW
AngleRateTmp = ((int32_t)(rate + 47) * rcCommand[YAW]) >> 5;
ROLL, PITCH
AngleRateTmp = ((int32_t)(rate + 27) * rcCommand[axis]) >> 4;
```

**Comment by martinbudden** [comment]
@ledvinap , wrt your suggestion about having a separate conversion function and new structure, I'm not really in favour of this, since the conversion is fairly straightforward. And the new structure will pretty much be the same as the pid profile.

**Comment by borisbstyle** [comment]
Let me have some thoughts about it

I think we could actually even introduce degrees/second numbers for rates.

That also requires some MSP/configurator changes, but is the way to go IMO.

Maybe not in this particular PR, but it is really easy to do it that way
and get rid of this rate scaling.

So rcRate parameter should be removed and only roll, pitch, yaw rate should
be used.

Also mw23 pid controller could be removed too and we can asd Mwi Super Expo
feature into rewrite and luxfloat. I already did this in betaflight with
success.
Op 8 apr. 2016 18:16 schreef "Martin Budden" <notifications@github.com>:

**Comment by martinbudden** [comment]
@borisbstyle I debated with myself about the line:
```
gyroRate = luxGyroScale * gyroADC[axis] * gyro.scale
```

The `gyro.scale` item here is set by the device driver and varies between gyrometers. `pidLuxFloat` is scaled to the gyro attached to the flight controller, `pidMultiWiiRewrite` is not, so how to square that circle?  My solution scales to the 3050, 6000, 6050 and 6500 gyros. It's not ideal, but it is a reasonable compromise.

**Comment by martinbudden** [comment]
@borisbstyle , yes, it would be good to introduce degrees/second values for rates. But this PR is getting complicated enough, so let's save that for a new PR.

**Comment by martinbudden** [comment]
@borisbstyle , I think there is a bug in your `filterApplyAverage` function in betaflight. You don't set `averageSum` to zero before you start summing.

**Comment by borisbstyle** [comment]
ok lets do that in seperate PR than.

I agree about keeping the scaling as it is. No function needed.

The scaling from all gyros is same. Its all 16.4f. That extra calculation
is not really needed.

Maybe we should convert rewrite also to deg/sec with integer pre scaler? It
makes it more readable I guess? Maybe in the next PR
Op 8 apr. 2016 18:29 schreef "Martin Budden" <notifications@github.com>:

**Comment by borisbstyle** [comment]
yes you are righ....averagesum should indeed be set to 0.

i do use seperate filtering as well additionally
Op 8 apr. 2016 18:36 schreef "Boris B." <dreambb1982@gmail.com>:

**Comment by martinbudden** [comment]
@borisbstyle , the scaling for the L3G4200 gyro is different. I don't know how widely it is used though.

**Comment by borisbstyle** [comment]
no targets out there what use it from what I know
Op 8 apr. 2016 18:46 schreef "Martin Budden" <notifications@github.com>:

**Comment by martinbudden** [comment]
@hydra , this is now ready for merge:

1. All comments considered, although not all acted upon
2. Flight tested
3. Test code passes
4. @borisbstyle has already copied scaling and rates into betaflight (although he has slightly increased yaw rates)

It now needs flight testing by a wider body of people. We may decide to increase the yaw rate after more people fly this, but that is a simple PR if required.

**Comment by borisbstyle** [comment]
@martinbudden 
I tested the scaling factors and I also did the equalize the rates between pid controllers and pretty happy with results.

**Comment by hydra** [comment]
@martinbudden thanks for rebase.

**Comment by hydra** [comment]
@martinbudden i like this since we get to delete a bunch of code.  will merge this once the build passes.

**Comment by hydra** [comment]
nice code size decrease too:

master:
```
   text	   data	    bss	    dec	    hex	filename
 129020	   1184	  19272	 149476	  247e4	./obj/main/cleanflight_SPRACINGF3MINI.elf
```

PR:

```
   text	   data	    bss	    dec	    hex	filename
 127344	   1168	  19160	 147672	  240d8	./obj/main/cleanflight_SPRACINGF3MINI.elf
```
reviewing now.

**Comment by hydra** [comment]
@martinbudden can you please review the documentation and update it accordingly via another PR?

**Comment by hydra** [comment]
@martinbudden can you also update the unit tests as you see fit.  and see spelling of `assess` vs `asses` :smile:   PR please.

**Comment by martinbudden** [comment]
@hydra , OK, I'll review the documentation.

As for the unit tests, this will allow them to be cleaned up significantly - the PIDs are much more testable now. I'll look into that as well.

File: src/main/flight/pid_luxfloat.c [hunk 3]
```diff
@@ -48,153 +49,159 @@
...
+        float delta = -(gyroRate - lastRateForDelta[axis]);
+        lastRateForDelta[axis] = gyroRate;
+        // Divide delta by dT to get differential (ie dr/dt)
+        delta *= (1.0f / dT);
+        if (pidProfile->dterm_cut_hz) {
+            // DTerm delta low pass filter
+            delta = applyBiQuadFilter(delta, &deltaFilterState[axis]);
+        } else {
+            // When DTerm low pass filter disabled apply moving average to reduce noise
+            delta = filterApplyAveragef(delta, DTERM_AVERAGE_COUNT, deltaState[axis]);
```
Comment (borisbstyle) [inline src/main/flight/pid_luxfloat.c]: Why not making average configurable?

**Comment by martinbudden** [comment]
Parsimony

**Comment by martinbudden** [comment]
@borisbstyle , sorry, I couldn't resist the one-word parsimonious reply. To properly answer your question:

1. Generally I think the default position should be not to be configurable, unless there is a demonstrated need.
2. Users can already configure the filter by configuring the `dterm_cut_hz` value
3. A moving average is not a very good low pass filter anyway, so why put lipstick on a pig?

**Comment by borisbstyle** [comment]
@martinbudden 
dterm_cut_hz is not enabled by default.

The options should be separated IMO. One can choose what to have....avarage, filtering or both.

Also the scaling doesnt match all scenarios now. Rewrite scaling is now decreased by removing that average summing. Rewrite D values will end up higher than current tunes.
I suggest changing this line in rewrite as this will give more close D values as before
delta = (delta * ((uint16_t)0xFFFF / ((uint16_t)targetLooptime >> 4))) >> 5;

and change luxfloat according that
static const float luxDTermScale = (0.000001f * (float)0xFFFF) / 512;


The above 2 will make sure that scaling will be same/simillar. Still slight difference due to a bit different gyro interpretation.

**Comment by borisbstyle** [comment]
I would also suggest to remove MW23 to have more consistency and always good defaults

**Comment by martinbudden** [comment]
@borisbstyle to address your points:

1. Would it be a good idea to enable `dterm_cut_hz` by default? If so could you suggest a default value?
2. Do people really use filtering and moving average together? Would not one or the other suffice?
3. On a related item, I note in betaflight you have changed to use a `filterApplyPt1` filter (you previously used `applyBiQuadFilter`). What was the reason for the change, and should Cleanflight also change?
4. The current DTerm scaling is consistent between LUX and MWREWRITE, but as you note, results in higher D values. I'll change this.
5. Whether or not to include MW23 is @hydra 's call. I don't have a preference either way. I do note that you took it out of betaflight and then put it back in - why was that?

**Comment by martinbudden** [comment]
@borisbstyle , don't you mean that `luxDTermScale` should change from
```
static const float luxDTermScale = (0.000001f * (float)0xFFFF) / 256;
```
to
```
static const float luxDTermScale = (0.000001f * (float)0xFFFF) / 128;
```

(not ` /512`)?

**Comment by martinbudden** [comment]
@Bengt-M here: https://github.com/cleanflight/cleanflight/blob/master/src/main/config/config_unittest.h#L70

that is within the macro `GET_PID_LUX_FLOAT_CORE_LOCALS`

**Pull request merged**